### ISSUE TITLE
Schema generation enhancements

### ISSUE THREAD

**Issue by dantownsend** [piccolo-orm/piccolo#193]
Piccolo now has a command for generating a schema from an existing database.

```bash
piccolo schema generate
```

The current implementation doesn't cover every possible edge case and database feature, so there's room for improvement. For example:

## Column defaults

Getting the defaults for a column, and reflecting them in the Piccolo column. For example, if the column has a default of `1` in the database, we want that to be reflected in the Piccolo schema as:

```python
class MyTable(Table):
    my_column = Integer(default=1)
```

## On Delete / On Update

The database values aren't currently reflected in the ``ForeignKey`` column definitions.

```python
class MyTable(Table):
    my_column = ForeignKey(on_delete=OnDelete.cascade)
```

## Decimal precision

The precision of a decimal column in the database currently isn't reflected in the ``Decimal`` column definitions.

```python
class MyTable(Table):
    my_column = Decimal(digits=(5,2))
```

--------

It's a fairly beginner friendly feature to work on, because even though the code is fairly advanced, it's completely self contained, and doesn't require extensive knowledge of the rest of Piccolo.

**Comment by wmshort** [comment]
I would be willing to take a crack at this, though I can't guarantee a quick turn-around.

**Comment by dantownsend** [comment]
@wmshort That would be great, thanks

**Comment by AliSayyah** [comment]
> WARNING: Unrecognised column types, added `Column` as a placeholder:
  test_table.value_updated ['double precision']

I think this problem can be added to this issue. unless it's the same thing as your `Decimal precision` part of the explanation.
and it says `Column` added as e placeholder. but these columns haven't been included in the generated code. they have been skipped.

**Comment by dantownsend** [comment]
@AliSayyah I'll create a separate issue for that - I think it's because Piccolo doesn't have a `double precision` column type, only `real`.

You're right though, even without supporting that column type, the schema generate command should still add a placeholder value. I'll create a separate issue for that too.

**Comment by AliSayyah** [comment]
> @AliSayyah I'll create a separate issue for that - I think it's because Piccolo doesn't have a `double precision` column type, only `real`.
> 
> You're right though, even without supporting that column type, the schema generate command should still add a placeholder value. I'll create a separate issue for that too.

That's good, thank you.
What is your opinion about adding support for views and materialized views?

**Comment by dantownsend** [comment]
@AliSayyah I think having support for views and materialised views would be a nice feature.

I haven't given much thought into what the API would look like though.

**Comment by AliSayyah** [comment]
@dantownsend Ok, so that's a thing to keep in mind for the future.

**Comment by wmshort** [comment]
Sorry for ignorance: if table-level triggers are give in information_schema.triggers, where are column-level triggers (ON DELETE etc) listed most readily?

**Comment by dantownsend** [comment]
@wmshort Good question - it looks like you can find `ON DELETE` and `ON UPDATE` here:

```
select * from information_schema.referential_constraints;
```

I found this gist on it:

https://gist.github.com/velosipedist/7250141

**Comment by wmshort** [comment]
@dantownsend Possible related bug? To test the trigger parsing, in ``example_app.py`` I changed one of ``Concert``'s ``Band`` foreign keys to trigger ``OnDelete.no_action``, however when querying the DB for the table's trigger information all I am getting back is ``CASCADE``. Could something be interfering in setting the correct ``OnDelete``/``OnUpdate`` values?

It seems like this might be the culprit?

``
# column_types.py
# This is here just for type inference - the actual value is set by
        # the Table metaclass. We can't set the actual value here, as
        # only the metaclass has access to the table.
        self._foreign_key_meta = ForeignKeyMeta(
            Table, OnDelete.cascade, OnUpdate.cascade
        )
``

**Comment by dantownsend** [comment]
@wmshort Yes, you're right - there seems to be a bug.

I'm glad you caught this.

I'll push a fix for it now.

**Comment by dantownsend** [comment]
@wmshort I've released a fix for this now - if you merge the latest master into your branch then hopefully it'll work.

Thanks for discovering that bug.

**Comment by wmshort** [comment]
@dantownsend Awesome, that should let me wrap up the other additions to schema generation (triggers, decimal precision). However with the merge of latest master the tests seem to have gotten borked

**Comment by dantownsend** [comment]
@wmshort Yeah, I think there's a couple of things going on.

The new columns in `tests/example_apps/music/tables.py`. Some of the existing tests would need updating to support the changes. We might be better off moving the new example tables into a separate file for now, so your tests and the existing tests can co-exist for now.

And there were some changes to `generate.py` in master which may have caused a bad merge.

**Comment by dantownsend** [comment]
@wmshort If I can help at all merging it in, let me know.

**Comment by wmshort** [comment]
@dantownsend  I think I've got generate.py sorted, but if you can manage a solution to the tests that would be fantastic. Then I can get the remaining additions in as separate PRs.

**Comment by dantownsend** [comment]
@wmshort No problem - will take a look.

**Comment by wmshort** [comment]
The last of these items is addressed by #276 , are there enhancements that could reasonably be added now? What is the status of the question of placeholders?

**Comment by dantownsend** [comment]
@wmshort Great, thanks for adding those. I'm not sure if any other features are missing - maybe indexes?

By placeholders, do you mean if a database column type isn't recognised by Piccolo?

**Comment by wmshort** [comment]
@dantownsend I'll look into the indexes, eefintiely! 

For the placeholders, does it make sense to include 'missing' field types? Would the generated schema actually be usable against an existing database? In any case there aren't many types still missingg so I will take your direction on this.

**Comment by dantownsend** [comment]
@wmshort There's a few really obscure column types in Postgres, and it'll be a while before we support them all (if at all):

https://www.postgresql.org/docs/current/datatype.html#DATATYPE-TABLE

I think it's possible to define custom column types in Postgres too, which complicates matters further.

You're right though, for a user who is dependent on an unsupported column type, it could be problematic.

I think there are only two solution - put a placeholder in there (just `Column`).

Or dynamically create a new column type on the fly:

```python
class SomeUnknownColumn(Column):
    pass
```

This is trickier though.

**Comment by wmshort** [comment]
I have index reflection working now, but it seems like this info was intentionally omitted through ``excluded columns`` etc. Was the choice for another reason than not (yet) having this introspection into the database?

**Comment by dantownsend** [comment]
@wmshort Yeah, it was just omitted because we didn't have the reflection capabilities at the time.
